python/tvm/autotvm/measure/measure_methods.py

```diff
@@ -496,7 +497,8 @@ def set_task(self, task):
 def _build_func_common(measure_input, runtime=None, check_gpu=None, build_option=None):
     """Common part for building a configuration"""
     target, task, config = measure_input
-    target, task.target_host = Target.check_and_update_host_consist(target, task.target_host)
+    assert not isinstance(target, (Map, dict)), "It's expected that 'target' is a string here."
```

Comment:
@Mousius I think we could but what would be the achievement exactly with a unit test for that internal function (`_build_func_common`)? More broadly, what has been our policy about testing `assert` in TVM code? Should we test all the asserts? I recall there were a discussion some time ago about it but could not find the PR :(

Reply:
OTOH, let me check if I can get rid of `Target.check_and_update_host_consist()` entirely.

Reply:
>  I think we could but what would be the achievement exactly with a unit test for that internal function (_build_func_common)?

@gromero it's generally good practice to have comprehensive test coverage, including input validation and error handling (which is essentially what these `assert`s do). One reason is to figure out whether or not we can actually trigger such a situation from a more public interface (i.e. test `LocalBuilder` in some way to trigger this), another is to document the expected interface of the function so it's well understood how we expect it to behave.

> More broadly, what has been our policy about testing assert in TVM code? Should we test all the asserts? I recall there were a discussion some time ago about it but could not find the PR, iirc you're discussing that with other Arm folks.

There was a discussion awhile back on https://github.com/apache/tvm/pull/9331#issuecomment-966241622, which was resolved by @manupa-arm merging the PR without the tests added, no follow up conversation has happened since. Different people follow the [Code Review Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) differently, and there are contributions without testing at all, so I think the policy is very much down to the Committer rather than strictly defined.

Reply:
> > I think we could but what would be the achievement exactly with a unit test for that internal function (_build_func_common)?
> 
> @gromero it's generally good practice to have comprehensive test coverage, including input validation and error handling (which is essentially what these `assert`s do). One reason is to figure out whether or not we can actually trigger such a situation from a more public interface (i.e. test `LocalBuilder` in some way to trigger this), another is to document the expected interface of the function so it's well understood how we expect it to behave.

I think my question was not clear, maybe I should have said "[...] unit test for that **assert** in that internal function [...]". I agreed with the test coverage necessity and also with Andrew's suggestion for adding the assert, and everything else about the assert. My question was actually about testing (with a unit test) the assert in that specific function.


> > More broadly, what has been our policy about testing assert in TVM code? Should we test all the asserts? I recall there were a discussion some time ago about it but could not find the PR, iirc you're discussing that with other Arm folks.
> 
> There was a discussion awhile back on [#9331 (comment)](https://github.com/apache/tvm/pull/9331#issuecomment-966241622), which was resolved by @manupa-arm merging the PR without the tests added, no follow up conversation has happened since. Different people follow the [Code Review Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) differently, and there are contributions without testing at all, so I think the policy is very much down to the Committer rather than strictly defined.

Thanks, that was exactly the PR I had in mind. OK, I've read it again carefully and I'm 100% with Manupa, Ashutosh, and Elen about we should not test internal asserts (or, as Elen says, "developer facing asserts" ). I totally see the necessary of having a good test coverage and adding tests like for https://github.com/apache/tvm/pull/10865, but for this case what is the value? Testing on a code refactor if the assert gets messed up or removed (a regression in the assert itself) from `_build_func_common`? Besides that the other tests for more public interfaces should also be already exercising the assert() properly (giving a more meaningful full traceback in case the assert fires).

Reply:
asserts aside, let me check again `check_and_update_host_consist`  and the whole Target implementation it might be better fix (there always is...). I think `check_and_update_host_consist` is just an helper (static) function added as helper for the transition for the new target object.

Reply:
> Thanks, that was exactly the PR I had in mind. OK, I've read it again carefully and I'm 100% with Manupa, Ashutosh, and Elen about we should not test internal asserts (or, as Elen says, "developer facing asserts" ). I totally see the necessary of having a good test coverage and adding tests like for https://github.com/apache/tvm/pull/10865, but for this case what is the value? 

We can also consider cases such as https://github.com/apache/tvm/pull/9299, where because we didn't introduce a test to ensure the behaviour, we opened ourselves up to typos which impact the behaviour without our knowledge - this was later fixed with error handling tested in https://github.com/apache/tvm/pull/9470. There's other instances such as https://github.com/apache/tvm/pull/9313 where there has been testing of internal functions error handling without any push back at all. Testing units in isolation is common practice in many software projects, they're low effort and quick to run so I'm unsure why there's such fierce pushback against them in TVM.

> Testing on a code refactor if the assert gets messed up or removed (a regression in the assert itself) from _build_func_common? Besides that the other tests for more public interfaces should also be already exercising the assert() properly (giving a more meaningful full traceback in case the assert fires).

This is an interesting point, because it raises the question as to where we should be testing this, I previously mentioned we might write a test in `LocalBuilder` to check this rather than actually exercising `_build_func_common` itself. I agree `_build_func_common` was not the best place to write the actual test for this behaviour but rather `check_and_update_host_consist`. `check_and_update_host_consist` itself doesn't have this as part of its interface (which seems the more appropriate place to do input validation) so @areusch had to reverse engineer it to figure out that interface, this seems to motivate testing internal interfaces. I'd rather see us asking how best to test something and ensure the internal interfaces are well understood than trying to find exclusions from our test coverage guidelines.

Reply:
I'd rather not leave the `assert` in to remember when we refactor `check_and_update_host_consist` given this likely is the wrong place for it and unless someone wants to check back on this PR it won't be clear why it's where it is. Can we leave the `assert` out here and rely on the refactor for `check_and_update_host_consist` to robustify it further down?

Reply:
@Mousius I spent some time looking to find the "highest-level" place we could put such an assert in the autotvm infrastructure. I believe it's [here](https://github.com/apache/tvm/blob/main/python/tvm/autotvm/tuner/tuner.py#L133). would that be better? We could just then and there assert that we always get an instance of `Target` (which should be true, looking at the log [decoder](https://github.com/apache/tvm/blob/main/python/tvm/autotvm/record.py#L157).

Reply:
Apologies for not getting back to you on this! I was referring to testing this at the higher interface than `_build_func_common` which I think you're right would be here as that's the `tune` interface. In doing so the assert could remain in `_build_func_common` as the interface would eventually be fulfilled by the internal function raising an error, but I think you're right that it'd be much clearer if the check was next to the inputs to `tune`.

Having said that, I would still strongly favour keeping the assertions inside of the `Target` API, as it seems as though we're defensively coding around our own internals here rather than actively wanting to clamp `tune` to only accepting these inputs?

Reply:
> Having said that, I would still strongly favour keeping the assertions inside of the Target API, as it seems as though we're defensively coding around our own internals here rather than actively wanting to clamp tune to only accepting these inputs?

i mean sure, i agree, but i think you're asking for a different PR to be written here. i think it'd be great to merge this 
(with the assert moved up to `tune`) as a step in the right direction of simplifying Target usage, and address the Target thing in a Target-focused PR rather than a cleanup in another component.

Reply:
I think we agree that the `assert` should be in the `Target` code, I totally agree that it's outside of the scope of this PR. 

If we introduce an `assert` in this PR, when does that get cleaned up? If it's an immediate follow up PR, then we can either not introduce it in this PR or we can put it anywhere as it'll get removed straight away. If not, we've left this in an untested state, with an extra `assert` for a different component which may never get cleaned up.

At a cursory glance I think `check_and_update_host_consist` supports any of the valid `Target` formats? So I'd suggest a follow up clarifying that with test cases.

In the spirit of getting this merged rather than continuing to debate that function here, I would go without the `assert` - what do you think?

Reply:
the reason i asked for the assert was that `build()` was previously called below passing `target_host=task.target`. that's different from what may happen now. when `target` is `Map`, Target.check_and_update_host_consist may [drop](https://github.com/apache/tvm/blob/main/python/tvm/target/target.py#L311) `task.target_host` from the returned `target` object. what's changing then is the way this particular function pre-processes `target`. I don't fully understand why someone would pass a `Map` in for `target` here, but if they did, this change could break that usage. i'd rather assert than silently break it. i don't care where we do it, but i do think avoiding silent break would be better. at some point we are going to get rid of `check_and_update_host_consist`, so putting the assert right next to it seems like the easiest way to prompt a future cleanup.